### ISSUE TITLE
ArgumentError (Invalid argument(s): argument value for 'return_value' is null)

### ISSUE THREAD

**Issue by mugikhan** [simolus3/sqlite3.dart#242]
I am getting an error in the latest version of Flutter (3.22.2) and sqlite3 (2.4.2). When checking for all exceptions, it seems to throw this exception `ArgumentError (Invalid argument(s): argument value for 'return_value' is null)` when closing the database connection. It also seems to occur on any query in the `sqlite3_step` function. The exception is not thrown in Flutter 3.19.6.

Any idea why the exception might be thrown despite the operations succeeding?

```dart
  int sqlite3_step(
    ffi.Pointer<sqlite3_stmt> pStmt,
  ) {
    return _sqlite3_step(
      pStmt,
    );
  }

  int sqlite3_close_v2(
    ffi.Pointer<sqlite3> db,
  ) {
    return _sqlite3_close_v2(
      db,
    );
  }
```

Flutter info
```
[✓] Flutter (Channel stable, 3.22.2, on macOS 14.5 23F79 darwin-arm64, locale en-ZA)
```

**Comment by simolus3** [comment]
This might be a regression in `dart:ffi` - thanks for the report. Do you have a stack trace or more details on how I could reproduce this. Do I need to somehow call `sqlite3_step` after closing? Do you have a complete stack trace on how this call happened?

**Comment by JCKodel** [comment]
Same here, using the PowerSync plugin (not using Drift):

<img width="867" alt="image" src="https://github.com/simolus3/sqlite3.dart/assets/379339/a9d80a34-c527-44a7-83e4-fd0b81b535c5">

<img width="867" alt="image" src="https://github.com/simolus3/sqlite3.dart/assets/379339/3cb29b63-b049-4f6d-b8d2-17301726a425">

The stacktrace is:

`sqlite3.g.dart`, line 59:

```dart
  int sqlite3_close_v2(
    ffi.Pointer<sqlite3> db,
  ) {
    return _sqlite3_close_v2(
      db,
    );
  }
```

`FfiBindings`, line 133:

```dart
  @override
  int sqlite3_close_v2() {
    return bindings.bindings.sqlite3_close_v2(db);
  }
```

`FinalizableDatabase`, line 39:

```dart
  @override
  void dispose() {
    for (final stmt in _statements) {
      stmt.dispose();
    }

    final code = database.sqlite3_close_v2();
    SqliteException? exception;
    if (code != SqlError.SQLITE_OK) {
      exception = createExceptionRaw(bindings, database, code,
          operation: 'closing database');
    }

    database.deallocateAdditionalMemory();

    if (exception != null) {
      throw exception;
    }
  }
}
```

`DatabaseImplementation`, line 232:

```dart
  @override
  void dispose() {
    if (_isClosed) return;

    disposeFinalizer.detach(this);
    _isClosed = true;

    for (final listener in _updateListeners) {
      listener.close();
    }
    database.sqlite3_update_hook(null);

    finalizable.dispose();
  }
```

`_PowerSyncDatabaseIsolateArgs`, line 546:

```dart
  rPort.listen((message) async {
    if (message is List) {
      String action = message[0];
      if (action == 'update') {
        updateController.add('update');
      } else if (action == 'close') {
        // This prevents any further transactions being opened, which would
        // eventually terminate the sync loop.
        await mutex.close();
        db?.dispose();
        db = null;
        updateController.close();
        upstreamDbClient.close();
        Isolate.current.kill();
      }
    }
  });
```

Basically, PowerSync calls `db.dispose()` when the sync isolate is done with its job.

I do not know if this is true for SQLite3, but SqliteF has some bugs on database close (because, apparently, Android always have a single connection to SQLite, and if you close it, all other instances using that database also closes, leading to executing queries on a closed database, because, for Android, `db` is a static singleton >.<)

**Comment by JCKodel** [comment]
BTW, this is also preventing compilation on Android:

```console
❯ flutter run --release
Launching lib/main.dart on moto g 6 plus in release mode...
../../../.pub-cache/hosted/pub.dev/sqlite3-2.4.3/lib/src/ffi/sqlite3.g.dart:56:7: Error: A non-null value must be returned since the return type 'int' doesn't allow null.
  int sqlite3_close_v2(
      ^
Target kernel_snapshot failed: Exception


FAILURE: Build failed with an exception.
```

**Comment by simolus3** [comment]
This appears to be a VM bug => https://github.com/dart-lang/sdk/issues/56064.

There is a workaround I could apply, but I'll wait to see if this will get fixed and backported.

**Comment by simolus3** [comment]
I've fixed this in version 2.4.4 of `package:sqlite3`. It's a workaround though, so if there's a cherry-pick/Dart stable that fixes this problem I'll revert it.
